Added example of from_tensors and from_tensor_slices

ashutosh1919 opened the pull request:
Added very important examples which differentiates `from_tensors` and `from_tensor_slices` in Dataset. The example describes the property of `from_tensor_slices` to merge the input tensor. In case of `from_tensors`, it doesn't merge input tensor.

ashutosh1919 committed:
Added example of from_tensors and from_tensor_slices

mihaimaruseac commented on tensorflow/python/data/ops/dataset_ops.py:
```diff
@@ -611,6 +618,13 @@ def from_tensor_slices(tensors):
            [b'B']], dtype=object))
 
+    >>> # `from_tensor_slices` merges the input tensor 
+    >>> # unlike `from_tensors` which will create 3D tensor in below example.
```
Merge these two lines in a single comment line

```
>>> # `from_tensor_slices` merges the input tensor, unlike `from_tensors` which increases dimensionality
```

(or two lines if it doesn't fit)

mihaimaruseac commented on tensorflow/python/data/ops/dataset_ops.py:
```diff
@@ -534,6 +534,13 @@ def from_tensors(tensors):
     [(array([1, 2, 3], dtype=int32), b'A')]
 
+    >>> # `from_tensors` creates 3D tensor in below example
+    >>> # unlike `from_tensor_slices` which merges the input tensor.
```
```
>>> # `from_tensors` adds one more dimension to the shape
>>> # use `from_tensor_slices` to merge the input tensor

ashutosh1919 committed:
Comments changed

aaudiber commented:
@ashutosh1919 Thank you for your interest in improving the docs!

I'm confused about what you mean by "merge the input tensor". Both methods convert their input argument to a `Tensor`. The only difference is whether they produce the whole tensor, or slices of the tensor. In particular, `dataset.from_tensor_slices(x)` is functionally equivalent to `dataset.from_tensors(x).unbatch()`. Does that clarify how things work?

ashutosh1919 commented:
> @ashutosh1919 Thank you for your interest in improving the docs!
> 
> I'm confused about what you mean by "merge the input tensor". Both methods convert their input argument to a `Tensor`. The only difference is whether they produce the whole tensor, or slices of the tensor. In particular, `dataset.from_tensor_slices(x)` is functionally equivalent to `dataset.from_tensors(x).unbatch()`. Does that clarify how things work?

@aaudiber , Thanks for the comment. Yes I understand that. Sorry for the language, I understand what it means but failed to describe it properly.

aaudiber reviewed (changes requested):
.

aaudiber commented:
@ashutosh1919 Can you share what behavior you originally expected, and how that was different from the actual behavior? Then we can figure out how to modify the doc to address the misunderstanding better

ashutosh1919 commented:
> @ashutosh1919 Can you share what behavior you originally expected, and how that was different from the actual behavior? Then we can figure out how to modify the doc to address the misunderstanding better

Sure. Please take a look at [this answer](https://stackoverflow.com/a/54051300/8289194)'s second part. I observed this behavior that `from_tensor_slices` slices the tensor along it's first dimension. But  `from_tensors` doesn't do that. Since, there was no example given in doc to explain that behavior difference in both, I have raised this PR.

aaudiber commented on tensorflow/python/data/ops/dataset_ops.py:
```diff
@@ -534,6 +534,13 @@ def from_tensors(tensors):
     >>> list(dataset.as_numpy_iterator())
     [(array([1, 2, 3], dtype=int32), b'A')]
 
+    >>> # `from_tensors` adds one more dimension to the shape
```
`from_tensors` doesn't change the shape of its input. The shape of
```[ tf.random_uniform([2, 3]), tf.random_uniform([2, 3]) ]```
 is `(2, 2, 3)`. If you pass a single `tf.random_uniform([2, 3])` instead of a list, the shape will be `(2, 3)`:

```
>>> dataset = tf.data.Dataset.from_tensors(tf.random_uniform([2, 3]))
>>> dataset.output_shapes
(2, 3)
```

I think what we might be missing here is explanation for when a user would want to use `from_tensors` as opposed to `from_tensor_slices`. The main use cases I can think of for using `from_tensors` are testing with a single example and creating artificial data, e.g. with `from_tensor_slices(example).repeat(1000)`.

ashutosh1919 replied:
Done

ashutosh1919 committed:
Comments changed

aaudiber reviewed (changes requested):
> Sure. Please take a look at [this answer](https://stackoverflow.com/a/54051300/8289194)'s second part. I observed this behavior that `from_tensor_slices` slices the tensor along it's first dimension. But `from_tensors` doesn't do that. Since, there was no example given in doc to explain that behavior difference in both, I have raised this PR.

Got it, thanks for the link! I think the potentially confusing part is actually in how `from_tensors` and `from_tensor_slices` interpret Python lists. In both cases, they merge the items of the list into a single tensor, but in the `from_tensor_slices` case we then slice that tensor back into a sequence of tensors.

ashutosh1919 committed:
Changed example and comments

ashutosh1919 commented:
@aaudiber , Thank you so much for clearing my doubts on this. I have fixed the comments and also changed the example as per your suggestions. I think now it makes more sense. Please review this one. Thanks again.

aaudiber commented on tensorflow/python/data/ops/dataset_ops.py:
```diff
@@ -534,6 +534,13 @@ def from_tensors(tensors):
     [(array([1, 2, 3], dtype=int32), b'A')]
 
+    >>> # `from_tensors` does not change the shape of input tensor.
+    >>> # use `from_tensor_slices` to slice the input tensor.
```
Can we put the comparison to `from_tensor_slices` in a sentence right after the first line? It's out of place here since the example code doesn't use `from_tensor_slices`. e.g.

```
"""Creates a `Dataset` with a single element, comprising the given tensors.

`from_tensors` produces a dataset containing only a single element. To slice the input 
tensor into multiple elements, use `from_tensor_slices` instead.
```

Instead we can describe the example according to what it does, e.g. "You can use `from_tensors` to produce a dataset which repeats the same example many times".

ashutosh1919 replied:
Yes changed it. Also changed `from_tensor_slices` example so that readers will be able to differentiate both methods in repeat() examples.

aaudiber commented on tensorflow/python/data/ops/dataset_ops.py:
```diff
@@ -534,6 +534,13 @@ def from_tensors(tensors):
+    >>> # use `from_tensor_slices` to slice the input tensor.
+    >>> example = tf.constant([1,2,3])
+    >>> dataset = tf.data.Dataset.from_tensors(example).repeat(1000)
+    >>> np.array(list(dataset1.as_numpy_iterator())).shape
```
This is a complicated way to print the shape. Can we simplify it to `dataset.element_spec`, or change the example so that we can print all elements of the dataset with `list(dataset.as_numpy_iterator())`?

ashutosh1919 replied:
done.

ashutosh1919 commented on tensorflow/python/data/ops/dataset_ops.py:
```diff
@@ -611,6 +618,13 @@ def from_tensor_slices(tensors):
+    >>> example = tf.constant([1,2,3])
+    >>> dataset = tf.data.Dataset.from_tensor_slices(example).repeat(1000)
+    >>> np.array(list(dataset1.as_numpy_iterator())).shape
+    (3000,)
```
This example doesn't add new information. But since I am adding example in `from_tensors`, I have also added similar example in `from_tensor_slices` so that one can differentiate both.

ashutosh1919 replied:
Changed it as described in below comment.

ashutosh1919 committed:
Changed examples and comments to show difference by repeating elements

ashutosh1919 commented:
@aaudiber , Changed examples and comments so that readers can differentiate both methods by looking at the results we get in two examples.

aaudiber commented on tensorflow/python/data/ops/dataset_ops.py:
```diff
@@ -611,6 +621,14 @@ def from_tensor_slices(tensors):
            [3, 2]], dtype=int32), array([[b'A'],
            [b'B']], dtype=object))
 
+    >>> # `from_tensor_slices` can also be used to repeat the examples
```
I think it's better to remove this section to keep the docstring focused. Each docstring should make sense on its own, but this relies on the user being familiar with the content of the `from_tensors` docstring

ashutosh1919 replied:
Done

aaudiber commented on tensorflow/python/data/ops/dataset_ops.py:
```diff
@@ -527,13 +527,23 @@ def _type_spec(self):
+    >>> example = tf.constant([1,2,3])
+    >>> dataset = tf.data.Dataset.from_tensors(example).repeat(2)
+    >>> list(dataset1.as_numpy_iterator())
+    [array([1, 2, 3], dtype=int32), array([1, 2, 3], dtype=int32)]
```
This section looks good to me :+1:

ashutosh1919 replied:
Done

ashutosh1919 committed:
Modified doc

ashutosh1919 commented:
@aaudiber , Removed section for `from_tensor_slices` and keeping other things as it is.

aaudiber reviewed (dismissed):
Thanks for the quick iterations, and for improving this doc!

gbaned commented:
@ashutosh1919 thank you, it is failing doctest can you please check here for [logs](https://source.cloud.google.com/results/invocations/27f15da0-a852-415d-9d83-a312ac980112/targets/%2F%2Ftensorflow%2Ftools%2Fdocs:tf_doctest/tests).

Please run the doctest locally as mentioned here in the [contributor guidelines](https://www.tensorflow.org/community/contribute/docs_ref).

ashutosh1919 committed:
Resolved doc error

ashutosh1919 commented:
@gbaned , @mihaimaruseac and @aaudiber , Sorry for that. You need to approve it again. There was slight problem in the doc which was failing tests.
